File: src/main/java/fr/raconteur/simpleskinswapper/gui/SkinWheelScreen.java
package fr.raconteur.simpleskinswapper.gui;

import fr.raconteur.simpleskinswapper.SimpleSkinSwapperClient;
import fr.raconteur.simpleskinswapper.changeskin.SkinChange;
import fr.raconteur.simpleskinswapper.changeskin.SkinSwapperState;
import java.util.*;
import net.minecraft.client.gui.GuiGraphicsExtractor;
import net.minecraft.client.gui.screens.Screen;
import net.minecraft.client.input.KeyEvent;
import net.minecraft.client.input.MouseButtonEvent;
import net.minecraft.core.ClientAsset;
import net.minecraft.network.chat.Component;
import net.minecraft.sounds.SoundEvents;
import net.minecraft.world.entity.player.PlayerModelType;
import net.minecraft.world.entity.player.PlayerSkin;

public class SkinWheelScreen extends Screen {

    private final Screen parent;
    private final List<SkinEntry> entries;
    private int selectedIndex = -1;

    private static final int MAX_ENTRIES = 10;
    private static final float OUTER_RADIUS = 90.0f;
    private static final float GAP_HALF_ANGLE = (float) Math.toRadians(2.0);

    private static final int COLOR_SECTOR       = 0xCC1A2535;
    private static final int COLOR_SECTOR_HOVER = 0xEE2B5F9E;
    private static final int COLOR_CENTER_BG    = 0xBB0D1627;
    private static final int COLOR_TEXT         = 0xFFFFFFFF;

    public SkinWheelScreen(Screen parent) {
        super(Component.empty());
        this.parent = parent;
        this.entries = loadWheelEntries();
    }

    @Override
    public boolean isPauseScreen() {
        return false;
    }

    //@Override
    //protected void renderBlurredBackground(GuiGraphicsExtractor context) {
    //    // No blur — wheel is a transparent overlay
    //}

    @Override
    public void extractBackground(GuiGraphicsExtractor context, int mouseX, int mouseY, float delta) {
        // No background — wheel is a transparent overlay
    }

    // -------------------------------------------------------------------------
    // Rendering
    // -------------------------------------------------------------------------

    @Override
    public void extractRenderState(GuiGraphicsExtractor context, int mouseX, int mouseY, float delta) {
        float cx = this.width / 2.0f;
        float cy = this.height / 2.0f;

        selectedIndex = getSelectedIndex(mouseX, mouseY, cx, cy);

        int n = entries.size();
        if (n == 0) {
            context.centeredText(font,
                    Component.translatable("simpleskinswapper.screen.carousel.no_skins"),
                    (int) cx, (int) cy, COLOR_TEXT);
            super.extractRenderState(context, mouseX, mouseY, delta);
            return;
        }

        // Draw pie sector backgrounds
        for (int i = 0; i < n; i++) {
            drawSector(context, cx, cy, i, n, i == selectedIndex);
        }

        // Center fill circle (on top of sectors)
        fillCircle(context, cx, cy, 28, COLOR_CENTER_BG);

        // Skin previews — painter's order: top (smallest py) first
        double sectorSize2 = 2 * Math.PI / n;
        double angleOffset2 = -Math.PI / 2 - sectorSize2 / 2.0;
        double previewDist2 = OUTER_RADIUS * 0.60;
        Integer[] order = new Integer[n];
        for (int i = 0; i < n; i++) order[i] = i;
        Arrays.sort(order, (a, b) -> {
            double pyA = cy + previewDist2 * Math.sin(angleOffset2 + sectorSize2 * a + sectorSize2 / 2.0);
            double pyB = cy + previewDist2 * Math.sin(angleOffset2 + sectorSize2 * b + sectorSize2 / 2.0);
            return Double.compare(pyA, pyB);
        });
        for (int i : order) {
            drawSectorPreview(context, cx, cy, i, n);
        }

        // Selected skin name above the wheel
        if (selectedIndex >= 0) {
            context.centeredText(font,
                    Component.nullToEmpty(entries.get(selectedIndex).displayName),
                    (int) cx, (int) (cy - OUTER_RADIUS) - font.lineHeight - 6, COLOR_TEXT);
        }

        super.extractRenderState(context, mouseX, mouseY, delta);
    }

    // -------------------------------------------------------------------------
    // Sector geometry
    // -------------------------------------------------------------------------

    private int getSelectedIndex(int mouseX, int mouseY, float cx, float cy) {
        double dx = mouseX - cx;
        double dy = mouseY - cy;
        double dist = Math.sqrt(dx * dx + dy * dy);
        if (dist < 10 || dist > OUTER_RADIUS * 1.1) return -1;

        int n = entries.size();
        double sectorSize = 2 * Math.PI / n;
        double angleOffset = -Math.PI / 2 - sectorSize / 2.0;

        double angle = Math.atan2(dy, dx);
        double adjusted = ((angle - angleOffset) % (2 * Math.PI) + 2 * Math.PI) % (2 * Math.PI);
        int idx = (int) (adjusted / sectorSize);
        return (idx >= 0 && idx < n) ? idx : -1;
    }

    private void drawSector(GuiGraphicsExtractor context, float cx, float cy, int index, int n, boolean hovered) {
        double sectorSize = 2 * Math.PI / n;
        double angleOffset = -Math.PI / 2 - sectorSize / 2.0;
        double baseAngle = angleOffset + sectorSize * index;
        double startAngle = baseAngle + GAP_HALF_ANGLE;
        double endAngle   = baseAngle + sectorSize - GAP_HALF_ANGLE;
        int color = hovered ? COLOR_SECTOR_HOVER : COLOR_SECTOR;
        fillSector(context, cx, cy, OUTER_RADIUS, startAngle, endAngle, color);
    }

    private void fillCircle(GuiGraphicsExtractor context, float cx, float cy, float radius, int color) {
        fillSector(context, cx, cy, radius, 0, 2 * Math.PI, color);
    }

    /**
     * Fills a pie sector using context.fill() — one call per pixel column.
     * Uses analytical cross-product tests to compute the y-range per column in O(1),
     * giving O(r) total instead of O(r²) with atan2.
     *
     * For a sector [startAngle, endAngle] with span < 2π:
     *   A point (dx, dy) is inside iff
     *     dy*cos(S) - dx*sin(S) >= 0   (left of start ray)
     *     dy*cos(E) - dx*sin(E) <= 0   (right of end ray)
     * Each constraint is linear in dy → gives yLo / yHi directly.
     */
    private void fillSector(GuiGraphicsExtractor context, float cx, float cy, float radius,
                            double startAngle, double endAngle, int color) {
        int r   = (int) Math.ceil(radius);
        int icx = (int) cx;
        int icy = (int) cy;

        double span = ((endAngle - startAngle) % (2 * Math.PI) + 2 * Math.PI) % (2 * Math.PI);
        if (span >= 2 * Math.PI - 1e-9) {
            // Full circle — skip angular tests entirely
            for (int dx = -r; dx <= r; dx++) {
                int dMax = (int) Math.sqrt(Math.max(0.0, radius * radius - (double) dx * dx));
                if (dMax > 0) context.fill(icx + dx, icy - dMax, icx + dx + 1, icy + dMax + 1, color);
            }
            return;
        }

        double cosS = Math.cos(startAngle), sinS = Math.sin(startAngle);
        double cosE = Math.cos(endAngle),   sinE = Math.sin(endAngle);

        for (int dx = -r; dx <= r; dx++) {
            double r2 = radius * radius - (double) dx * dx;
            if (r2 <= 0) continue;
            int dyMax = (int) Math.sqrt(r2);
            double yLo = -dyMax, yHi = dyMax;

            // Start-ray constraint: dy*cosS - dx*sinS >= 0  →  dy >= dx*sinS/cosS
            double tS = (double) dx * sinS;
            if      (cosS >  1e-9) yLo = Math.max(yLo, tS / cosS);
            else if (cosS < -1e-9) yHi = Math.min(yHi, tS / cosS);
            else if (tS   >  1e-9) continue; // column entirely outside start ray

            // End-ray constraint: dy*cosE - dx*sinE <= 0  →  dy <= dx*sinE/cosE
            double tE = (double) dx * sinE;
            if      (cosE >  1e-9) yHi = Math.min(yHi, tE / cosE);
            else if (cosE < -1e-9) yLo = Math.max(yLo, tE / cosE);
            else if (tE   < -1e-9) continue; // column entirely outside end ray

            int fillY1 = (int) Math.ceil(yLo);
            int fillY2 = (int) Math.floor(yHi);
            if (fillY1 <= fillY2) {
                context.fill(icx + dx, icy + fillY1, icx + dx + 1, icy + fillY2 + 1, color);
            }
        }
    }

    private void drawSectorPreview(GuiGraphicsExtractor context, float cx, float cy, int index, int n) {
        double sectorSize = 2 * Math.PI / n;
        double angleOffset = -Math.PI / 2 - sectorSize / 2.0;
        double midAngle = angleOffset + sectorSize * index + sectorSize / 2.0;
        double previewDist = OUTER_RADIUS * 0.60;

        int px = (int) (cx + previewDist * Math.cos(midAngle));
        int py = (int) (cy + previewDist * Math.sin(midAngle));

        SkinEntry entry = entries.get(index);
        entry.ensureTextureLoaded();

        int halfW = 16;
        int halfH = 24;

        if (entry.textureId != null) {
            PlayerSkin skinTextures = new PlayerSkin(
                    new ClientAsset.DownloadedTexture(entry.textureId, ""), null, null,
                    entry.skinType == SkinType.SLIM ? PlayerModelType.SLIM : PlayerModelType.WIDE,
                    true
            );
            SkinRenderer.renderPlayer(context, px - halfW, py - halfH, px + halfW, py + halfH, halfH, skinTextures);
        }
    }

    // -------------------------------------------------------------------------
    // Input handling
    // -------------------------------------------------------------------------

    @Override
    public boolean mouseClicked(MouseButtonEvent click, boolean doubled) {
        if (click.button() == 0) { apply(); return true; }
        if (click.button() == 1) { onClose(); return true; }
        return super.mouseClicked(click, doubled);
    }

    @Override
    public boolean keyReleased(KeyEvent input) {
        if (SimpleSkinSwapperClient.openWheelKey.matches(input)) {
            onClose();
            return true;
        }
        return super.keyReleased(input);
    }

    private void apply() {
        if (selectedIndex >= 0 && selectedIndex < entries.size()) {
            SkinEntry entry = entries.get(selectedIndex);
            if (SkinSwapperState.beginSwap()) {
                minecraft.player.playSound(SoundEvents.UI_BUTTON_CLICK.value(), 1.0f, 1.0f);
                SkinChange.changeSkin(
                        entry.file,
                        entry.skinType,
                        entry.textureId,
                        () -> {
                            if (minecraft.player != null)
                                minecraft.player.sendOverlayMessage(
                                        Component.translatable("simpleskinswapper.message.success"));
                        },
                        err -> {
                            if (minecraft.player != null)
                                minecraft.player.sendOverlayMessage(
                                        Component.translatable("simpleskinswapper.message.error", err));
                        }
                );
                if (minecraft.player != null)
                    minecraft.player.sendOverlayMessage(
                            Component.translatable("simpleskinswapper.message.applying"));
            }
        }
    }

    @Override
    public void onClose() {
        minecraft.gui.setScreen(parent);
    }

    // -------------------------------------------------------------------------
    // Data loading — reads order.txt, no writes
    // -------------------------------------------------------------------------

    private static List<SkinEntry> loadWheelEntries() {
        List<SkinEntry> ordered = SkinCarouselScreen.loadOrderedEntries();
        if (ordered.size() <= MAX_ENTRIES) return ordered;
        return new ArrayList<>(ordered.subList(0, MAX_ENTRIES));
    }
}

File: src/main/kotlin/fr/raconteur/simpleskinswapper/gui/SkinWheelScreen.kt
package fr.raconteur.simpleskinswapper.gui

import com.mojang.blaze3d.platform.InputConstants
import fr.raconteur.simpleskinswapper.SimpleSkinSwapperClient
import fr.raconteur.simpleskinswapper.overlayMessage
import fr.raconteur.simpleskinswapper.changeskin.SkinChange
import fr.raconteur.simpleskinswapper.changeskin.SkinSwapperState
import net.minecraft.client.gui.GuiGraphicsExtractor
import net.minecraft.client.gui.screens.Screen
import net.minecraft.client.input.KeyEvent
import net.minecraft.client.input.MouseButtonEvent
import net.minecraft.core.ClientAsset
import net.minecraft.network.chat.Component
import net.minecraft.resources.Identifier
import net.minecraft.sounds.SoundEvents
import net.minecraft.util.Mth
import net.minecraft.world.entity.player.PlayerModelType
import net.minecraft.world.entity.player.PlayerSkin
import org.joml.Matrix3x2f

class SkinWheelScreen(private val parent: Screen?) : Screen(Component.empty()) {

    private val entries: List<SkinEntry> = loadWheelEntries()
    private var selectedIndex = -1
    private val hoverAnimFactors = FloatArray(entries.size)
    private var lastHoverAnimUpdateNanos = 0L

    override fun isPauseScreen(): Boolean = false

    //override fun renderBlurredBackground(context: GuiGraphicsExtractor) {
    //    // No blur — wheel is a transparent overlay
    //}

    override fun extractBackground(context: GuiGraphicsExtractor, mouseX: Int, mouseY: Int, delta: Float) {
        // No background — wheel is a transparent overlay
    }

    // -------------------------------------------------------------------------
    // Rendering
    // -------------------------------------------------------------------------

    //? if >=26.1 {
    override fun extractRenderState(context: GuiGraphicsExtractor, mouseX: Int, mouseY: Int, delta: Float) {
    //?} else {
    /*override fun render(context: GuiGraphicsExtractor, mouseX: Int, mouseY: Int, delta: Float) {
    *///?}
        val cx = this.width / 2.0f
        val cy = this.height / 2.0f

        selectedIndex = getSelectedIndex(mouseX, mouseY, cx, cy)

        val n = entries.size
        if (n == 0) {
            context.centeredText(
                font,
                Component.translatable("simpleskinswapper.screen.carousel.no_skins"),
                cx.toInt(), cy.toInt(), COLOR_TEXT
            )
            //? if >=26.1 {
            //? if >=26.1 {
        super.extractRenderState(context, mouseX, mouseY, delta)
        //?} else {
        /*super.render(context, mouseX, mouseY, delta)
        *///?}
            //?} else {
            /*super.render(context, mouseX, mouseY, delta)
            *///?}
            return
        }

        // Draw pie sector backgrounds
        for (i in 0..<n) {
            drawSector(context, cx, cy, i, n, i == selectedIndex)
        }

        // Center fill circle (on top of sectors)
        fillCircle(context, cx, cy, 28f, COLOR_CENTER_BG)

        updateHoverAnimations()

        // Skin previews — painter's order: top (smallest py) first
        val sectorSize2 = 2 * Math.PI / n
        val angleOffset2 = -Math.PI / 2 - sectorSize2 / 2.0
        val previewDist2 = OUTER_RADIUS * 0.60
        val order = Array(n) { it }
        order.sortBy { i ->
            cy + previewDist2 * Math.sin(angleOffset2 + sectorSize2 * i + sectorSize2 / 2.0)
        }
        for (i in order) {
            drawSectorPreview(context, cx, cy, i, n)
        }

        // Selected skin name above the wheel
        if (selectedIndex >= 0) {
            context.centeredText(
                font,
                Component.nullToEmpty(entries[selectedIndex].displayName),
                cx.toInt(), (cy - OUTER_RADIUS).toInt() - font.lineHeight - 6, COLOR_TEXT
            )
        }

        //? if >=26.1 {
        super.extractRenderState(context, mouseX, mouseY, delta)
        //?} else {
        /*super.render(context, mouseX, mouseY, delta)
        *///?}
    }

    // -------------------------------------------------------------------------
    // Sector geometry
    // -------------------------------------------------------------------------

    private fun getSelectedIndex(mouseX: Int, mouseY: Int, cx: Float, cy: Float): Int {
        val dx = mouseX - cx
        val dy = mouseY - cy
        val dist = Math.sqrt((dx * dx + dy * dy).toDouble())
        if (dist < 10 || dist > OUTER_RADIUS * 1.1) return -1

        val n = entries.size
        val sectorSize = 2 * Math.PI / n
        val angleOffset = -Math.PI / 2 - sectorSize / 2.0

        val angle = Math.atan2(dy.toDouble(), dx.toDouble())
        val adjusted = ((angle - angleOffset) % (2 * Math.PI) + 2 * Math.PI) % (2 * Math.PI)
        val idx = (adjusted / sectorSize).toInt()
        return if (idx >= 0 && idx < n) idx else -1
    }

    private fun drawSector(context: GuiGraphicsExtractor, cx: Float, cy: Float, index: Int, n: Int, hovered: Boolean) {
        val sectorSize = 2 * Math.PI / n
        val angleOffset = -Math.PI / 2 - sectorSize / 2.0
        val baseAngle = angleOffset + sectorSize * index
        val startAngle = baseAngle + GAP_HALF_ANGLE
        val endAngle = baseAngle + sectorSize - GAP_HALF_ANGLE
        val color = if (hovered) COLOR_SECTOR_HOVER else COLOR_SECTOR
        submitSectorFill(context, cx, cy, OUTER_RADIUS, startAngle.toFloat(), endAngle.toFloat(), color)
    }

    private fun fillCircle(context: GuiGraphicsExtractor, cx: Float, cy: Float, radius: Float, color: Int) {
        submitSectorFill(context, cx, cy, radius, 0.0f, (2 * Math.PI).toFloat(), color)
    }

    /** Submits one sector as a single triangle-fan mesh — O(1) draw submissions per sector. */
    private fun submitSectorFill(
        context: GuiGraphicsExtractor, cx: Float, cy: Float, radius: Float,
        startAngle: Float, endAngle: Float, color: Int
    ) {
        context.guiRenderState.addGuiElement(
            SectorFillRenderState(
                Matrix3x2f(context.pose()), cx, cy, radius, startAngle, endAngle, color,
                context.scissorStack.peek()
            )
        )
    }

    /** Eases every sector's hover factor toward 1 (hovered) or 0 with an exponential settle. */
    private fun updateHoverAnimations() {
        val now = System.nanoTime()
        val dt = if (lastHoverAnimUpdateNanos == 0L) 0.0F else (now - lastHoverAnimUpdateNanos) / 1_000_000_000.0F
        lastHoverAnimUpdateNanos = now
        val t = 1.0F - Math.exp((-HOVER_ANIM_SPEED * dt).toDouble()).toFloat()
        for (i in hoverAnimFactors.indices) {
            val target = if (i == selectedIndex) 1.0F else 0.0F
            var eased = Mth.lerp(t, hoverAnimFactors[i], target)
            if (Math.abs(eased - target) < HOVER_ANIM_SNAP_EPSILON) eased = target
            hoverAnimFactors[i] = eased
        }
    }

    private fun drawSectorPreview(context: GuiGraphicsExtractor, cx: Float, cy: Float, index: Int, n: Int) {
        val sectorSize = 2 * Math.PI / n
        val angleOffset = -Math.PI / 2 - sectorSize / 2.0
        val midAngle = angleOffset + sectorSize * index + sectorSize / 2.0
        val previewDist = OUTER_RADIUS * 0.60

        val px = (cx + previewDist * Math.cos(midAngle)).toInt()
        val py = (cy + previewDist * Math.sin(midAngle)).toInt()

        val entry = entries[index]
        entry.ensureTextureLoaded()

        val halfW = 16
        val halfH = 24

        val textureId = entry.textureId ?: return

        // Every preview is a live entity render; only the hovered one plays the walk animation.
        SkinRenderer.renderPlayer(
            context, px - halfW, py - halfH, px + halfW, py + halfH, halfH,
            buildPlayerSkin(entry, textureId), hoverAnimFactors[index]
        )
    }

    private fun buildPlayerSkin(entry: SkinEntry, textureId: Identifier): PlayerSkin =
        PlayerSkin(
            ClientAsset.DownloadedTexture(textureId, ""), null, null,
            if (entry.skinType == SkinType.SLIM) PlayerModelType.SLIM else PlayerModelType.WIDE,
            true
        )

    // -------------------------------------------------------------------------
    // Input handling
    // -------------------------------------------------------------------------

    override fun mouseClicked(click: MouseButtonEvent, doubled: Boolean): Boolean {
        // Button codes follow the platform: GLFW numbering (left=0) on <=26.2, SDL (left=1) on 26.3+.
        if (click.button() == InputConstants.MOUSE_BUTTON_LEFT) {
            apply()
            return true
        }
        if (click.button() == InputConstants.MOUSE_BUTTON_RIGHT) {
            onClose()
            return true
        }
        return super.mouseClicked(click, doubled)
    }

    override fun keyReleased(input: KeyEvent): Boolean {
        if (SimpleSkinSwapperClient.openWheelKey?.matches(input) == true) {
            onClose()
            return true
        }
        return super.keyReleased(input)
    }

    private fun apply() {
        if (selectedIndex >= 0 && selectedIndex < entries.size) {
            val entry = entries[selectedIndex]
            if (SkinSwapperState.beginSwap()) {
                minecraft.player?.playSound(SoundEvents.UI_BUTTON_CLICK.value(), 1.0f, 1.0f)
                SkinChange.changeSkin(
                    entry.file,
                    entry.skinType,
                    entry.textureId,
                    {
                        minecraft.player?.overlayMessage(
                            Component.translatable("simpleskinswapper.message.success")
                        )
                    },
                    { err ->
                        minecraft.player?.overlayMessage(
                            Component.translatable("simpleskinswapper.message.error", err)
                        )
                    }
                )
                minecraft.player?.overlayMessage(
                    Component.translatable("simpleskinswapper.message.applying")
                )
            }
        }
    }

    override fun onClose() {
        //? if >=26.2 {
        minecraft.gui.setScreen(parent)
        //?} else {
        /*minecraft.setScreen(parent)
        *///?}
    }

    companion object {
        private const val MAX_ENTRIES = 10
        private const val OUTER_RADIUS = 90.0f
        private val GAP_HALF_ANGLE = Math.toRadians(2.0).toFloat()

        // Hover animation easing: higher = faster settle back to the neutral pose.
        private const val HOVER_ANIM_SPEED = 10.0F
        private const val HOVER_ANIM_SNAP_EPSILON = 0.05F

        private val COLOR_SECTOR = 0xCC1A2535.toInt()
        private val COLOR_SECTOR_HOVER = 0xEE2B5F9E.toInt()
        private val COLOR_CENTER_BG = 0xBB0D1627.toInt()
        private val COLOR_TEXT = 0xFFFFFFFF.toInt()

        // -------------------------------------------------------------------------
        // Data loading — reads order.txt, no writes
        // -------------------------------------------------------------------------

        private fun loadWheelEntries(): List<SkinEntry> {
            val ordered = SkinCarouselScreen.loadOrderedEntries()
            if (ordered.size <= MAX_ENTRIES) return ordered
            return ArrayList(ordered.subList(0, MAX_ENTRIES))
        }
    }
}
